### PR TITLE
Feature : AlephDNS

### PR DESCRIPTION
**Pull request by 1yam**
Implement AlephDNS from @aliel inside SDK

### CODE REVIEW

File: src/aleph/sdk/conf.py [hunk 1]
```diff
@@ -33,6 +33,12 @@ class Settings(BaseSettings):
 
     CODE_USES_SQUASHFS: bool = which("mksquashfs") is not None  # True if command exists
 
+    # Dns resolver
+    IPFS_DOMAINS = "ipfs.public.aleph.sh"
```
Comment (odesenfans) [inline src/aleph/sdk/conf.py:37]: Why `DOMAINS`, plural?

File: src/aleph/sdk/conf.py [hunk 2]
```diff
@@ -33,6 +33,12 @@ class Settings(BaseSettings):
...
+    PROGRAM_DOMAINS = "program.public.aleph.sh"
+    ROOT_DOMAIN = "static.public.aleph.sh"
+    RESOLVERS = ["1.1.1.1", "8.8.8.8"]
```
Comment (odesenfans) [inline src/aleph/sdk/conf.py:40]: `DNS_RESOLVERS`? Also, using Google's DNS by default is not so web3.

File: tests/unit/test_domains.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+"""
+@pytest.mark.asyncio
+async def test_cname():
+    alephdns = AlephDNS()
+    url = 'https://custom_domain_test.aleph.sh'
+    check = await alephdns.custom_domain_check(url)
+    assert check is not None
+"""
```
Comment (odesenfans) [inline tests/unit/test_domains.py:42]: Why is this test commented?

File: src/aleph/sdk/domain.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            raise DomainConfigurationError(error)
+
+    async def check_domain(self, url: str, _type: str, owner: Optional[str] = None):
+        # if _type.lower() == 'ipfs':
+        return await self.check_ipfs_domain(url, _type, owner)
+        # elif _type.lower() == 'program':
+        #    pass
+
+    async def check_ipfs_domain(
+        self, url: str, _type: str, owner: Optional[str] = None
```
Comment (odesenfans) [inline src/aleph/sdk/domain.py:58]: Having a variable prefixed with an underscore in a function makes very little sense, unless it's a parameter that's not supposed to be used for common use cases.

File: src/aleph/sdk/domain.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            print("Check...", _type)
+            return await self.check_domain(domain, _type, owner)
+        except Exception as error:
+            raise DomainConfigurationError(error)
+
+    async def check_domain(self, url: str, _type: str, owner: Optional[str] = None):
+        # if _type.lower() == 'ipfs':
+        return await self.check_ipfs_domain(url, _type, owner)
+        # elif _type.lower() == 'program':
+        #    pass
```
Comment (odesenfans) [inline src/aleph/sdk/domain.py:55]: Remove commented code.

File: src/aleph/sdk/domain.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            expected_value = f"_dnslink.{domain}.{settings.ROOT_DOMAIN}"
+            assert_error = (
+                f"CNAME record not found: _dnslink.{domain}",
+                f"Create a CNAME record for _dnslink.{domain} with value: {expected_value}",
+                status,
+            )
+
+            assert res is not None, assert_error
+            assert hasattr(res, "cname"), assert_error
+            assert res.cname == expected_value, assert_error
```
Comment (odesenfans) [inline src/aleph/sdk/domain.py:106]: Lots of asserts. If the conditions are not supposed to happen at all, fine. If they can happen at runtime, these should be converted to proper exceptions (`assert` can be deactivated in Python).

**Review by odesenfans — DISMISSED** [review]

**Review by aliel — COMMENTED** [review]
:ok_hand:

**Comment by MHHukiewitz** [comment]
Merge? Yes/No/Only with PBJ?

**Review by aliel — APPROVED** [review]
to Merge

File: src/aleph/sdk/domain.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+class AlephDNS:
+    def __init__(self):
+        self.resolver = aiodns.DNSResolver(servers=settings.DNS_RESOLVERS)
+        self.fqdn_matcher = re.compile(r"https?://?")
+
+    async def query(self, name: str, query_type: str):
+        try:
+            return await self.resolver.query(name, query_type)
+        except Exception as e:
```
Comment (hoh) [inline src/aleph/sdk/domain.py:16]: @aliel what is the exception that we are expected here ? Catching all exceptions to stdout is not a best practice.

File: src/aleph/sdk/domain.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+import re
+from .conf import settings
+from typing import Optional
+from aleph.sdk.exceptions import DomainConfigurationError
+
+
+class AlephDNS:
+    def __init__(self):
+        self.resolver = aiodns.DNSResolver(servers=settings.DNS_RESOLVERS)
+        self.fqdn_matcher = re.compile(r"https?://?")
```
Comment (hoh) [inline src/aleph/sdk/domain.py:11]: Why the second `?` at the end ?

File: src/aleph/sdk/domain.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    async def query(self, name: str, query_type: str):
+        try:
+            return await self.resolver.query(name, query_type)
+        except Exception as e:
+            print(e)
+            return None
+
+    def url_to_domain(self, url):
+        return self.fqdn_matcher.sub("", url).strip().strip("/")
```
Comment (hoh) [inline src/aleph/sdk/domain.py:21]: You can simplify the `strip()` call
```python
return self.fqdn_matcher.sub("", url).strip(" /")
```

File: src/aleph/sdk/domain.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+    async def get_ipv6_address(self, url: str):
```
Comment (hoh) [inline src/aleph/sdk/domain.py:23]: What type does this return ?

File: src/aleph/sdk/domain.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+    async def get_dnslink(self, url: str):
+        domain = self.url_to_domain(url)
+        query = await self.query(f"_dnslink.{domain}", "TXT")
+        if query is not None and len(query) > 0:
+            return query[0].text
+
+    async def check_domain_configured(self, domain, target, owner):
+        try:
+            print("Check...", target)
```
Comment (hoh) [inline src/aleph/sdk/domain.py:40]: Why not use `logging.debug` instead of `print` here ?

File: src/aleph/sdk/domain.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        dns_rules.append({
+            "rule_name": "cname",
+            "dns": {
+                "type": "cname",
+                "name": domain,
+                "value": cname_value
+            },
+            "info": f"Create a CNAME record for {domain} with value {cname_value}",
+            "on_error": f"CNAME record not found: {domain}"
+        })
```
Comment (hoh) [inline src/aleph/sdk/domain.py:107]: Would it make sense to use a dataclasses instead of a dict for this data ?

File: src/aleph/sdk/domain.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    async def check_domain_configured(self, domain, target, owner):
+        try:
+            print("Check...", target)
+            return await self.check_domain(domain, target, owner)
+        except Exception as error:
+            raise DomainConfigurationError(error)
+
+    async def check_domain(
+        self, url: str, target: str, owner: Optional[str] = None
+    ):
```
Comment (hoh) [inline src/aleph/sdk/domain.py:47]: Can you add a docstring to explain what this is checking ?

File: src/aleph/sdk/domain.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+                found = False
+
+                for _res in res:
+                    if hasattr(_res, "text") and _res.text == record_value:
+                        found = True
+
+                if found == False:
+                    raise DomainConfigurationError(
+                        (dns_rule["info"], dns_rule["on_error"], status)
+                    )
```
Comment (hoh) [inline src/aleph/sdk/domain.py:74]: Why not use a `for ... else ...` loop ?

```python
                for _res in res:
                    if hasattr(_res, "text") and _res.text == record_value:
                        break
                else:
                    raise DomainConfigurationError(
                        (dns_rule["info"], dns_rule["on_error"], status)
                    )

**Review by hoh — COMMENTED** [review]
Sorry for reviewing late for merge, but I have a few issues with the quality of this branch that should be improved before releasing it.

**Pull request merged**